src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/ScreenRecorder.cs

```diff
@@ -100,18 +103,92 @@ private void OnDestroy()
 #endif
             if (_isRecording)
             {
-                KeyboardInputActionObserver.GetInstance()?.StopRecording();
-                MouseInputActionObserver.GetInstance()?.StopRecording();
-                var theVideoDirectory = _currentVideoDirectory;
-                Task.Run(() =>
+                RGBotManager.GetInstance().ShowUploadingIndicator(true);
+                HandleRecordingSaving();
+            }
+        }
+
+        /**
+         * We separate out the async saving code since OnDestroy can't be async, and we need to wait for saving
+         * before doing API calls
+         */
+        private async void HandleRecordingSaving()
+        {
+            _isRecording = false;
+            var endTime = DateTime.Now;
+            KeyboardInputActionObserver.GetInstance()?.StopRecording();
+            MouseInputActionObserver.GetInstance()?.StopRecording();
+            var gameplaySessionDataDirectory = _currentGameplaySessionDirectoryPrefix + "_data";
+            var gameplaySessionScreenshotDirectory = _currentGameplaySessionDirectoryPrefix + "_screenshots";
+            var thumbnailPath = _currentGameplaySessionDirectoryPrefix + "_thumbnail.png";
+            await Task.Run(() =>
```

Comment:
Running this on another thread but awaiting it seems pointless , why not just run it where we are